rfcs/0019-data-structures/0019-data-structures.md

```diff
@@ -0,0 +1,255 @@
+---
+Number: "0000"
+Category: Standards Track
+Status: Proposal
+Author: Haichao Zhu, Xuejie Xiao
+Organization: Nervos Foundation
+Created: 2019-03-26
+---
+
+# Data Structures of Nervos CKB
+
+This documents explains all the basic data structures used in CKB.
+
+* [Cell](#Cell)
+* [Script](#Script)
+* [Transaction](#Transaction)
+* [Block](#Block)
+
+
+
+## Cell
+
+### Example
+
+```json
+{
+    "capacity": 5000000,
+    "data": "0x",
+    "lock": {
+      "args": [],
+      "binary_hash": "0xa58a960b28d6e283546e38740e80142da94f88e88d5114d8dc91312b8da4765a",
+      "version": 0
+    },
+    "type": null
+}
+```
+
+## Description
+
+| Name       | Type       | Description                                                  |
+| :--------- | :--------- | :----------------------------------------------------------- |
+| `capacity` | uint64     | **The size of the cell.** When a new cell is generated (via transaction), one of the verification rule is `capacity ≥ len(capacity)+len(data)+len(type)+len(lock)`. This value also represents the balance of CKB coin, just like the `nValue` field in the Bitcoin's CTxOut. (E.g. Alice owns 100 CKB coins means she can unlock a group of cells that has 100 amount of `capacity` in total.) |
+| `data`     | Bytes      | **Arbitrary data.** This part is for storing states or scripts.  In order to make this cell valid on-chain, the data filled in this field should comply with the logics and rules defined by `type`. |
+| `type`     | `Script`   | **A Script that defines the type of the cell.** It limits how the `data` field of the new cells can be changed from old cells. `type` is required to has a data structure of `script`. **This field is optional.** |
+| `lock`     | `Script`   | **A Script that defines the ownership of the cell**, just like the `scriptPubKey` field in the Bitcoin's CTxOut. Whoever can provide unlock arguments that makes the execution of this script success can consume this cell as input in an transaction (i.e. has the ownership of this cell). |
+
+
+
+More information about Cell can be found in the [whitepaper](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0002-ckb/0002-ckb.md#42-cell).
+
+
+
+## Script
+
+### Example
+
+```json
+{
+  "version": 0,
```

Comment:
`version` will be removed from `Script` cc @xxuejie

Reply:
Yes, but what about we removed this once the PR at CKB is crafted?

Reply:
That's ok. We can merge this PR after CKB PR. No hurry.